improvement(parser): size suffixes in quota creation parameters

Eridanus117 opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
add support for size suffixes in quota creation statement parameters

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=49087) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=49087) before we can accept your contribution.<br/><hr/>**Eridanus** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=49087) it.</sub>

Enmk commented on src/Parsers/Access/ParserCreateQuotaQuery.cpp:
```diff
@@ -107,19 +108,27 @@ namespace
+    T fieldToNumber(const Field & f)
+    {
+        if (f.getType() == Field::Types::String)
+            return parseWithSizeSuffix<T>(f.get<const String &>());
```
How do you handle error input here? Please add some negative-case tests, like:
- `query_inserts = '1!'`
- `query_inserts = '1test'`
- etc

ucasfl replied:
Exception will be throwed, well it's better not to throw exception in parser.

Enmk replied:
IMO it is not that bad... CH already throws exceptions from parser code. BUT still having an error-case test would be really nice.
```
src/Parsers$ rgrep --include 'Parser*.cpp' -Han throw
ParserCreateQuery.cpp:687:            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Multiple primary keys are not allowed.");
ParserSetQuery.cpp:260:            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Parameter name cannot be empty");
ParserInsertQuery.cpp:201:            throw Exception(ErrorCodes::SYNTAX_ERROR,
ParserInsertQuery.cpp:231:            throw Exception(ErrorCodes::SYNTAX_ERROR, "You have excessive ';' symbol before data for INSERT.\n"
ParserTablesInSelectQuery.cpp:193:                throw Exception(ErrorCodes::SYNTAX_ERROR, "You must not specify ANY or ALL for CROSS JOIN.");
ParserTablesInSelectQuery.cpp:197:                throw Exception(ErrorCodes::SYNTAX_ERROR, "SEMI|ANTI JOIN should be LEFT or RIGHT.");
ParserExternalDDLQuery.cpp:65:                throw Exception(ErrorCodes::MYSQL_SYNTAX_ERROR, "Cannot parse MySQL alter query.");
ParserExternalDDLQuery.cpp:68:                throw Exception(ErrorCodes::MYSQL_SYNTAX_ERROR, "Cannot parse MySQL rename query.");
ParserExternalDDLQuery.cpp:71:                throw Exception(ErrorCodes::MYSQL_SYNTAX_ERROR, "Cannot parse MySQL drop query.");
ParserExternalDDLQuery.cpp:74:                throw Exception(ErrorCodes::MYSQL_SYNTAX_ERROR, "Cannot parse MySQL create query.");
ParserSelectQuery.cpp:327:            throw Exception(ErrorCodes::LIMIT_BY_WITH_TIES_IS_NOT_SUPPORTED, "Can not use WITH TIES alongside LIMIT BY/DISTINCT ON");
ParserSelectQuery.cpp:335:                throw Exception(ErrorCodes::LIMIT_BY_WITH_TIES_IS_NOT_SUPPORTED, "Can not use WITH TIES alongside LIMIT BY/DISTINCT ON");
ParserSelectQuery.cpp:338:                throw Exception(ErrorCodes::SYNTAX_ERROR, "Can not use DISTINCT ON alongside LIMIT BY");
ParserSelectQuery.cpp:350:            throw Exception(ErrorCodes::TOP_AND_LIMIT_TOGETHER, "Can not use TOP and LIMIT together");
ParserSelectQuery.cpp:363:                throw Exception(ErrorCodes::ROW_AND_ROWS_TOGETHER, "Can not use ROW and ROWS together");
ParserSelectQuery.cpp:375:                throw Exception(ErrorCodes::OFFSET_FETCH_WITHOUT_ORDER_BY, "Can not use OFFSET FETCH clause without ORDER BY");
ParserSelectQuery.cpp:380:                    throw Exception(ErrorCodes::FIRST_AND_NEXT_TOGETHER, "Can not use FIRST and NEXT together");
ParserSelectQuery.cpp:391:                    throw Exception(ErrorCodes::ROW_AND_ROWS_TOGETHER, "Can not use ROW and ROWS together");
ParserSelectQuery.cpp:455:        throw Exception(ErrorCodes::WITH_TIES_WITHOUT_ORDER_BY, "Can not use WITH TIES without ORDER BY");
Kusto/ParserKQLOperators.cpp:23:        throw Exception(ErrorCodes::SYNTAX_ERROR, "Syntax error near {}", kql_op);
Kusto/ParserKQLOperators.cpp:58:        throw Exception(ErrorCodes::SYNTAX_ERROR, "Syntax error near {}", kql_op);
Kusto/ParserKQLOperators.cpp:118:        throw Exception(ErrorCodes::SYNTAX_ERROR, "Syntax error near {}", kql_op);
Kusto/ParserKQLOperators.cpp:138:                throw Exception(ErrorCodes::SYNTAX_ERROR, "Invalid negative operator");
Access/ParserCreateUserQuery.cpp:189:                        throw Exception(ErrorCodes::OPENSSL_ERROR, "Cannot generate salt for password. OpenSSL {}", buf);
Access/ParserCreateUserQuery.cpp:203:                    ///if USE_SSL is not defined, Exception thrown later
Access/ParserGrantQuery.cpp:159:    void throwIfNotGrantable(AccessRightsElements & elements)
Access/ParserGrantQuery.cpp:171:                throw Exception(ErrorCodes::INVALID_GRANT, "{} cannot be granted on the column level", old_flags.toString());
Access/ParserGrantQuery.cpp:173:                throw Exception(ErrorCodes::INVALID_GRANT, "{} cannot be granted on the table level", old_flags.toString());
Access/ParserGrantQuery.cpp:175:                throw Exception(ErrorCodes::INVALID_GRANT, "{} cannot be granted on the database level", old_flags.toString());
Access/ParserGrantQuery.cpp:177:                throw Exception(ErrorCodes::INVALID_GRANT, "{} cannot be granted", old_flags.toString());
Access/ParserGrantQuery.cpp:284:        throw Exception(ErrorCodes::SYNTAX_ERROR, "GRANT OPTION should be specified for access types");
Access/ParserGrantQuery.cpp:286:        throw Exception(ErrorCodes::SYNTAX_ERROR, "ADMIN OPTION should be specified for roles");
Access/ParserGrantQuery.cpp:310:            throwIfNotGrantable(elements);
Access/ParserShowCreateAccessEntityQuery.cpp:153:            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Type {} is not implemented in SHOW CREATE query", toString(type));
Access/ParserCreateQuotaQuery.cpp:74:            throw Exception(ErrorCodes::SYNTAX_ERROR, "Quota cannot be keyed by '{}'. Expected one of the following identifiers: {}", name, all_types_str);
```

Enmk commented on tests/queries/0_stateless/01297_create_quota.sql:
IMO it is not clear what value was set exactly, especially for erroneous input, like invalid string or for something that may cause integer overflow/underflow. E.g. :
- `query_selects = '184467440737 T'`
- `query_selects = '-1'`
- `query_selects = 'foobar'`

Eridanus117 committed:
improvement(parser): size suffixes in quota creation parameters

This commit adds support for creating quotas with size suffixes in the
parameter list of a quota creation statement. The parser now recognizes
size suffixes such as "KB", "MB", and "GB" and correctly converts them
to their corresponding values in bytes. This feature improves the
usability of the quota creation statement and makes it more flexible.
The commit also includes additional tests to ensure that the new syntax
is parsed correctly.

ucasfl commented on tests/queries/0_stateless/01297_create_quota.sql:
```diff
@@ -137,3 +138,76 @@ DROP QUOTA q1_01297, q2_01297;
+SELECT '-- bad syntax test';
+CREATE QUOTA q1_01297 FOR INTERVAL 1 MINUTE MAX query_selects = '1 '; -- { clientError 62 }
+CREATE QUOTA q2_01297 FOR INTERVAL 1 MINUTE MAX execution_time = '1 '; -- { clientError 62 }
+CREATE QUOTA q3_01297 FOR INTERVAL 1 MINUTE MAX query_selects = '1K '; -- { clientError 62 }
```
I think this case can be supported.

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #49087 from Eridanus117/ClickHouse:suffix_size -->
This is an automated comment for commit d479404318a6c29a9c02b498989d6cdf44343b82 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49087/d479404318a6c29a9c02b498989d6cdf44343b82/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td>🔴 failure</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
<tr><td>Push to Dockerhub</td><td>The check for building and pushing the CI related docker images to docker hub</td><td>🟢 success</td></tr>
<tr><td>Run Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>🟡 pending</td></tr>
<tr><td>Style Check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td>🟢 success</td></tr>
</table>

Eridanus117 committed:
empty commit

Eridanus117 committed:
Merge branch 'master' into suffix_size

Eridanus117 committed:
Merge branch 'master' into suffix_size

alexey-milovidov commented on src/Parsers/Access/ParserCreateQuotaQuery.cpp:
```diff
@@ -107,23 +107,182 @@ namespace
     }
 
+    template <typename T, typename = std::enable_if_t<std::is_same_v<T, QuotaValue> || std::is_same_v<T, double>>>
+    bool multipleBinarySuffix(T & max_value, int power_of_two)
```
It's strange that we have suffix-parsing code here while we already support it for Settings.

alexey-milovidov replied:
ReadHelpers.h, `parseWithSizeSuffix`

Eridanus117 committed:
improvement(parser): add support for size suffixes in quota creation
statement parameters

This commit adds support for creating quotas with size suffixes in the
parameter list of a quota creation statement. The parser now recognizes
size suffixes such as "KB", "MB", and "GB" and correctly converts them
to their corresponding values in bytes. This feature improves the
usability of the quota creation statement and makes it more flexible.
The commit also includes additional tests to ensure that the new syntax
is parsed correctly.
Attention: we did not support string literal with double like
execute_time = '1.5'

Eridanus117 committed:
improvement(parser): add support for size suffixes in quota creation
statement parameters

This commit adds support for creating quotas with size suffixes in the
parameter list of a quota creation statement. The parser now recognizes
size suffixes such as "KB", "MB", and "GB" and correctly converts them
to their corresponding values in bytes. This feature improves the
usability of the quota creation statement and makes it more flexible.
The commit also includes additional tests to ensure that the new syntax
is parsed correctly.
Attention: we did not support string literal with double like
execute_time = '1.5'

Eridanus117 committed:
test(parser): add hint for create quota overflow test

Eridanus117 committed:
test(parser): remove confusing quota overflow test

alexey-milovidov reviewed (approved):
All good to me.
While I doubt if trimming the whitespace is needed, it's ok.
